// @walletconnect/client/dist/cjs/controllers/pairing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pairing = void 0;
const tslib_1 = require("tslib");
const events_1 = require("events");
const pino_utils_1 = require("@pedrouid/pino-utils");
const types_1 = require("@walletconnect/types");
const utils_1 = require("@walletconnect/utils");
const utils_2 = require("@json-rpc-tools/utils");
const subscription_1 = require("./subscription");
const history_1 = require("./history");
const constants_1 = require("../constants");
class Pairing extends types_1.IPairing {
    constructor(client, logger) {
        super(client, logger);
        this.client = client;
        this.logger = logger;
        this.events = new events_1.EventEmitter();
        this.context = constants_1.PAIRING_CONTEXT;
        this.logger = pino_utils_1.generateChildLogger(logger, this.context);
        this.pending = new subscription_1.Subscription(client, this.logger, constants_1.PAIRING_STATUS.pending, false);
        this.settled = new subscription_1.Subscription(client, this.logger, constants_1.PAIRING_STATUS.settled, true);
        this.history = new history_1.JsonRpcHistory(client, this.logger);
        this.registerEventListeners();
    }
    init() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.trace(`Initialized`);
            yield this.pending.init();
            yield this.settled.init();
            yield this.history.init();
        });
    }
    get(topic) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.settled.get(topic);
        });
    }
    ping(topic) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const request = { method: constants_1.PAIRING_JSONRPC.ping, params: {} };
            return this.request({ topic, request });
        });
    }
    send(topic, payload) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const pairing = yield this.settled.get(topic);
            const encryptKeys = {
                sharedKey: pairing.sharedKey,
                publicKey: pairing.self.publicKey,
            };
            if (utils_2.isJsonRpcRequest(payload)) {
                if (!Object.values(constants_1.PAIRING_JSONRPC).includes(payload.method)) {
                    if (!pairing.permissions.jsonrpc.methods.includes(payload.method)) {
                        const errorMessage = `Unauthorized JSON-RPC Method Requested: ${payload.method}`;
                        this.logger.error(errorMessage);
                        throw new Error(errorMessage);
                    }
                    yield this.history.set(topic, payload);
                    payload = utils_2.formatJsonRpcRequest(constants_1.PAIRING_JSONRPC.payload, {
                        request: { method: payload.method, params: payload.params },
                    }, payload.id);
                }
            }
            else {
                yield this.history.update(topic, payload);
            }
            yield this.client.relayer.publish(pairing.topic, payload, {
                relay: pairing.relay,
                encryptKeys,
            });
        });
    }
    get length() {
        return this.settled.length;
    }
    get topics() {
        return this.settled.topics;
    }
    get values() {
        return this.settled.values.map(x => x.data);
    }
    create(params) {
        return new Promise((resolve, reject) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.debug(`Create Pairing`);
            this.logger.trace({ type: "method", method: "create", params });
            const timeout = setTimeout(() => {
                const errorMessage = `Pairing failed to settle after 30 seconds`;
                this.logger.error(errorMessage);
                reject(errorMessage);
            }, 30000);
            let pending;
            try {
                pending = yield this.propose(params);
            }
            catch (e) {
                clearTimeout(timeout);
                return reject(e);
            }
            this.pending.on(constants_1.SUBSCRIPTION_EVENTS.updated, (updatedEvent) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                if (pending.topic !== updatedEvent.data.topic)
                    return;
                if (utils_1.isPairingResponded(updatedEvent.data)) {
                    const outcome = updatedEvent.data.outcome;
                    clearTimeout(timeout);
                    if (utils_1.isPairingFailed(outcome)) {
                        try {
                            yield this.pending.delete(pending.topic, outcome.reason);
                        }
                        catch (e) {
                            return reject(e);
                        }
                        reject(new Error(outcome.reason));
                    }
                    else {
                        try {
                            const pairing = yield this.settled.get(outcome.topic);
                            yield this.pending.delete(pending.topic, constants_1.PAIRING_REASONS.settled);
                            resolve(pairing);
                        }
                        catch (e) {
                            return reject(e);
                        }
                    }
                }
            }));
        }));
    }
    respond(params) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.debug(`Respond Pairing`);
            this.logger.trace({ type: "method", method: "respond", params });
            const { approved, proposal } = params;
            const self = utils_1.generateKeyPair();
            if (approved) {
                try {
                    const responder = {
                        publicKey: self.publicKey,
                    };
                    const expiry = Date.now() + proposal.ttl;
                    const pairing = yield this.settle({
                        relay: proposal.relay,
                        self,
                        peer: proposal.proposer,
                        permissions: proposal.permissions,
                        ttl: proposal.ttl,
                        expiry,
                    });
                    const outcome = {
                        topic: pairing.topic,
                        relay: pairing.relay,
                        responder,
                        expiry,
                    };
                    const pending = {
                        status: constants_1.PAIRING_STATUS.responded,
                        topic: proposal.topic,
                        relay: proposal.relay,
                        self,
                        proposal,
                        outcome,
                    };
                    yield this.pending.set(pending.topic, pending, { relay: pending.relay });
                    return pending;
                }
                catch (e) {
                    const reason = e.message;
                    const outcome = { reason };
                    const pending = {
                        status: constants_1.PAIRING_STATUS.responded,
                        topic: proposal.topic,
                        relay: proposal.relay,
                        self,
                        proposal,
                        outcome,
                    };
                    yield this.pending.set(pending.topic, pending, { relay: pending.relay });
                    return pending;
                }
            }
            else {
                const outcome = { reason: constants_1.PAIRING_REASONS.not_approved };
                const pending = {
                    status: constants_1.PAIRING_STATUS.responded,
                    topic: proposal.topic,
                    relay: proposal.relay,
                    self,
                    proposal,
                    outcome,
                };
                yield this.pending.set(pending.topic, pending, { relay: pending.relay });
                return pending;
            }
        });
    }
    update(params) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.debug(`Update Pairing`);
            this.logger.trace({ type: "method", method: "update", params });
            const pairing = yield this.settled.get(params.topic);
            const participant = { publicKey: pairing.self.publicKey };
            const update = yield this.handleUpdate(pairing, params, participant);
            const request = utils_2.formatJsonRpcRequest(constants_1.PAIRING_JSONRPC.update, update);
            yield this.send(pairing.topic, request);
            return pairing;
        });
    }
    request(params) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                const request = utils_2.formatJsonRpcRequest(params.request.method, params.request.params);
                const timeout = setTimeout(() => {
                    const errorMessage = `JSON-RPC Request timeout after 30s: ${request.method}`;
                    this.logger.error(errorMessage);
                    reject(errorMessage);
                }, 30000);
                this.events.on(constants_1.PAIRING_EVENTS.payload, (payloadEvent) => {
                    if (params.topic !== payloadEvent.topic)
                        return;
                    if (utils_2.isJsonRpcRequest(payloadEvent.payload))
                        return;
                    const response = payloadEvent.payload;
                    if (response.id !== request.id)
                        return;
                    clearTimeout(timeout);
                    if (utils_2.isJsonRpcError(response)) {
                        const errorMessage = response.error.message;
                        this.logger.error(errorMessage);
                        return reject(new Error(errorMessage));
                    }
                    return resolve(response.result);
                });
                try {
                    yield this.send(params.topic, request);
                }
                catch (e) {
                    clearTimeout(timeout);
                    return reject(e);
                }
            }));
        });
    }
    delete(params) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.debug(`Delete Pairing`);
            this.logger.trace({ type: "method", method: "delete", params });
            yield this.settled.delete(params.topic, params.reason);
        });
    }
    on(event, listener) {
        this.events.on(event, listener);
    }
    once(event, listener) {
        this.events.once(event, listener);
    }
    off(event, listener) {
        this.events.off(event, listener);
    }
    removeListener(event, listener) {
        this.events.removeListener(event, listener);
    }
    propose(params) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.debug(`Propose Pairing`);
            this.logger.trace({ type: "method", method: "propose", params });
            const relay = (params === null || params === void 0 ? void 0 : params.relay) || { protocol: constants_1.RELAYER_DEFAULT_PROTOCOL };
            const topic = utils_1.generateRandomBytes32();
            const self = utils_1.generateKeyPair();
            const proposer = {
                publicKey: self.publicKey,
            };
            const uri = utils_1.formatUri({
                protocol: this.client.protocol,
                version: this.client.version,
                topic: topic,
                publicKey: proposer.publicKey,
                relay: relay,
            });
            const permissions = {
                jsonrpc: {
                    methods: [constants_1.SESSION_JSONRPC.propose],
                },
            };
            const proposal = {
                relay,
                topic,
                proposer,
                signal: {
                    method: constants_1.PAIRING_SIGNAL_METHOD_URI,
                    params: { uri },
                },
                permissions,
                ttl: constants_1.PAIRING_DEFAULT_TTL,
            };
            const pending = {
                status: constants_1.PAIRING_STATUS.proposed,
                topic: proposal.topic,
                relay: proposal.relay,
                self,
                proposal,
            };
            yield this.pending.set(pending.topic, pending, { relay });
            return pending;
        });
    }
    settle(params) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.logger.debug(`Settle Pairing`);
            this.logger.trace({ type: "method", method: "settle", params });
            const sharedKey = utils_1.deriveSharedKey(params.self.privateKey, params.peer.publicKey);
            const topic = yield utils_1.sha256(sharedKey);
            const pairing = {
                topic,
                relay: params.relay,
                sharedKey,
                self: params.self,
                peer: params.peer,
                permissions: params.permissions,
                expiry: params.expiry,
            };
            const decryptKeys = {
                sharedKey,
            };
            yield this.settled.set(pairing.topic, pairing, { relay: pairing.relay, decryptKeys });
            return pairing;
        });
    }
    onResponse(payloadEvent) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { topic, payload } = payloadEvent;
            this.logger.debug(`Receiving Pairing response`);
            this.logger.trace({ type: "method", method: "onResponse", topic, payload });
            const request = payload;
            const pending = yield this.pending.get(topic);
            let errorMessage;
            if (!utils_1.isPairingFailed(request.params)) {
                try {
                    const pairing = yield this.settle({
                        relay: pending.relay,
                        self: pending.self,
                        peer: request.params.responder,
                        permissions: pending.proposal.permissions,
                        ttl: pending.proposal.ttl,
                        expiry: request.params.expiry,
                    });
                    yield this.pending.update(topic, {
                        status: constants_1.PAIRING_STATUS.responded,
                        outcome: {
                            topic: pairing.topic,
                            relay: pairing.relay,
                            responder: request.params.responder,
                            expiry: pairing.expiry,
                        },
                    });
                }
                catch (e) {
                    this.logger.error(e);
                    errorMessage = e.message;
                    yield this.pending.update(topic, {
                        status: constants_1.PAIRING_STATUS.responded,
                        outcome: { reason: e.message },
                    });
                }
                const response = typeof errorMessage === "undefined"
                    ? utils_2.formatJsonRpcResult(request.id, true)
                    : utils_2.formatJsonRpcError(request.id, errorMessage);
                yield this.client.relayer.publish(pending.topic, response, { relay: pending.relay });
            }
            else {
                this.logger.error(request.params.reason);
                yield this.pending.update(topic, {
                    status: constants_1.PAIRING_STATUS.responded,
                    outcome: { reason: request.params.reason },
                });
            }
        });
    }
    onAcknowledge(payloadEvent) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { topic, payload } = payloadEvent;
            this.logger.debug(`Receiving Pairing acknowledge`);
            this.logger.trace({ type: "method", method: "onAcknowledge", topic, payload });
            const response = payload;
            const pending = yield this.pending.get(topic);
            if (!utils_1.isPairingResponded(pending))
                return;
            if (utils_2.isJsonRpcError(response) && !utils_1.isPairingFailed(pending.outcome)) {
                yield this.settled.delete(pending.outcome.topic, response.error.message);
            }
            yield this.pending.delete(topic, constants_1.PAIRING_REASONS.acknowledged);
        });
    }
    onMessage(payloadEvent) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { topic, payload } = payloadEvent;
            this.logger.debug(`Receiving Pairing message`);
            this.logger.trace({ type: "method", method: "onMessage", topic, payload });
            if (utils_2.isJsonRpcRequest(payload)) {
                const request = payload;
                const pairing = yield this.settled.get(payloadEvent.topic);
                let errorMessage = "";
                switch (request.method) {
                    case constants_1.PAIRING_JSONRPC.payload:
                        yield this.onPayload(payloadEvent);
                        break;
                    case constants_1.PAIRING_JSONRPC.update:
                        yield this.onUpdate(payloadEvent);
                        break;
                    case constants_1.PAIRING_JSONRPC.delete:
                        yield this.settled.delete(pairing.topic, request.params.reason);
                        break;
                    case constants_1.PAIRING_JSONRPC.ping:
                        yield this.send(pairing.topic, utils_2.formatJsonRpcResult(request.id, false));
                        break;
                    default:
                        errorMessage = `Unknown JSON-RPC Method Requested: ${request.method}`;
                        this.logger.error(errorMessage);
                        yield this.send(pairing.topic, utils_2.formatJsonRpcError(request.id, errorMessage));
                        break;
                }
            }
            else {
                this.onPayloadEvent(payloadEvent);
            }
        });
    }
    onPayload(payloadEvent) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { topic, payload } = payloadEvent;
            if (utils_2.isJsonRpcRequest(payload)) {
                const { id, params } = payload;
                const request = utils_2.formatJsonRpcRequest(params.request.method, params.request.params, id);
                const pairing = yield this.settled.get(topic);
                if (!pairing.permissions.jsonrpc.methods.includes(request.method)) {
                    const errorMessage = `Unauthorized JSON-RPC Method Requested: ${request.method}`;
                    this.logger.error(errorMessage);
                    throw new Error(errorMessage);
                }
                const pairingPayloadEvent = {
                    topic,
                    payload: request,
                };
                this.logger.debug(`Receiving Pairing payload`);
                this.logger.trace(Object.assign({ type: "method", method: "onPayload" }, pairingPayloadEvent));
                this.onPayloadEvent(pairingPayloadEvent);
            }
            else {
                const pairingPayloadEvent = {
                    topic,
                    payload,
                };
                this.logger.debug(`Receiving Pairing payload`);
                this.logger.trace(Object.assign({ type: "method", method: "onPayload" }, pairingPayloadEvent));
                this.onPayloadEvent(pairingPayloadEvent);
            }
        });
    }
    onUpdate(payloadEvent) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { topic, payload } = payloadEvent;
            this.logger.debug(`Receiving Pairing update`);
            this.logger.trace({ type: "method", method: "onUpdate", topic, payload });
            const request = payloadEvent.payload;
            const pairing = yield this.settled.get(payloadEvent.topic);
            try {
                const participant = { publicKey: pairing.peer.publicKey };
                yield this.handleUpdate(pairing, { topic, update: request.params }, participant);
                const response = utils_2.formatJsonRpcResult(request.id, true);
                yield this.send(pairing.topic, response);
            }
            catch (e) {
                this.logger.error(e);
                const response = utils_2.formatJsonRpcError(request.id, e.message);
                yield this.send(pairing.topic, response);
            }
        });
    }
    handleUpdate(pairing, params, participant) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            let update;
            if (typeof params.update.peer !== "undefined") {
                const metadata = params.update.peer.metadata;
                pairing.peer.metadata = metadata;
                update = { peer: { metadata } };
            }
            else {
                const errorMessage = `Invalid pairing update request params`;
                this.logger.error(errorMessage);
                throw new Error(errorMessage);
            }
            if (participant.publicKey === pairing.self.publicKey) {
                yield this.settled.update(pairing.topic, pairing);
            }
            return update;
        });
    }
    shouldIgnorePayloadEvent(payloadEvent) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { topic, payload } = payloadEvent;
            if (!this.settled.subscriptions.has(topic))
                return true;
            let exists = false;
            try {
                exists = yield this.history.exists(topic, payload.id);
            }
            catch (e) {
            }
            return exists;
        });
    }
    onPayloadEvent(payloadEvent) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { topic, payload } = payloadEvent;
            if (utils_2.isJsonRpcRequest(payload)) {
                if (yield this.shouldIgnorePayloadEvent(payloadEvent))
                    return;
                yield this.history.set(topic, payload);
            }
            else {
                yield this.history.update(topic, payload);
            }
            this.logger.info(`Emitting ${constants_1.PAIRING_EVENTS.payload}`);
            this.logger.debug({ type: "event", event: constants_1.PAIRING_EVENTS.payload, data: payloadEvent });
            this.events.emit(constants_1.PAIRING_EVENTS.payload, payloadEvent);
        });
    }
    onPendingPayloadEvent(event) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (utils_2.isJsonRpcRequest(event.payload)) {
                switch (event.payload.method) {
                    case constants_1.PAIRING_JSONRPC.approve:
                    case constants_1.PAIRING_JSONRPC.reject:
                        this.onResponse(event);
                        break;
                    default:
                        break;
                }
            }
            else {
                this.onAcknowledge(event);
            }
        });
    }
    onPendingStatusEvent(event) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const pending = event.data;
            if (utils_1.isPairingResponded(pending)) {
                this.logger.info(`Emitting ${constants_1.PAIRING_EVENTS.responded}`);
                this.logger.debug({ type: "event", event: constants_1.PAIRING_EVENTS.responded, data: pending });
                this.events.emit(constants_1.PAIRING_EVENTS.responded, pending);
                if (!utils_1.isSubscriptionUpdatedEvent(event)) {
                    const method = !utils_1.isPairingFailed(pending.outcome)
                        ? constants_1.PAIRING_JSONRPC.approve
                        : constants_1.PAIRING_JSONRPC.reject;
                    const request = utils_2.formatJsonRpcRequest(method, pending.outcome);
                    yield this.client.relayer.publish(pending.topic, request, { relay: pending.relay });
                }
            }
            else {
                this.logger.info(`Emitting ${constants_1.PAIRING_EVENTS.proposed}`);
                this.logger.debug({ type: "event", event: constants_1.PAIRING_EVENTS.proposed, data: pending });
                this.events.emit(constants_1.PAIRING_EVENTS.proposed, pending);
            }
        });
    }
    registerEventListeners() {
        this.pending.on(constants_1.SUBSCRIPTION_EVENTS.payload, (payloadEvent) => this.onPendingPayloadEvent(payloadEvent));
        this.pending.on(constants_1.SUBSCRIPTION_EVENTS.created, (createdEvent) => this.onPendingStatusEvent(createdEvent));
        this.pending.on(constants_1.SUBSCRIPTION_EVENTS.updated, (updatedEvent) => this.onPendingStatusEvent(updatedEvent));
        this.settled.on(constants_1.SUBSCRIPTION_EVENTS.payload, (payloadEvent) => this.onMessage(payloadEvent));
        this.settled.on(constants_1.SUBSCRIPTION_EVENTS.created, (createdEvent) => {
            const pairing = createdEvent.data;
            this.logger.info(`Emitting ${constants_1.PAIRING_EVENTS.settled}`);
            this.logger.debug({ type: "event", event: constants_1.PAIRING_EVENTS.settled, data: pairing });
            this.events.emit(constants_1.PAIRING_EVENTS.settled, pairing);
        });
        this.settled.on(constants_1.SUBSCRIPTION_EVENTS.updated, (updatedEvent) => {
            const pairing = updatedEvent.data;
            this.logger.info(`Emitting ${constants_1.PAIRING_EVENTS.updated}`);
            this.logger.debug({ type: "event", event: constants_1.PAIRING_EVENTS.updated, data: pairing });
            this.events.emit(constants_1.PAIRING_EVENTS.updated, pairing);
        });
        this.settled.on(constants_1.SUBSCRIPTION_EVENTS.deleted, (deletedEvent) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            const pairing = deletedEvent.data;
            this.logger.info(`Emitting ${constants_1.PAIRING_EVENTS.deleted}`);
            this.logger.debug({ type: "event", event: constants_1.PAIRING_EVENTS.deleted, data: pairing });
            this.events.emit(constants_1.PAIRING_EVENTS.deleted, pairing);
            const request = utils_2.formatJsonRpcRequest(constants_1.PAIRING_JSONRPC.delete, {
                reason: deletedEvent.reason,
            });
            yield this.history.delete(pairing.topic);
            yield this.client.relayer.publish(pairing.topic, request, { relay: pairing.relay });
        }));
    }
}
exports.Pairing = Pairing;
