// @walletconnect/client/dist/cjs/constants/pairing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PAIRING_REASONS = exports.PAIRING_EVENTS = exports.PAIRING_STATUS = exports.PAIRING_SIGNAL_METHOD_URI = exports.PAIRING_DEFAULT_TTL = exports.PAIRING_CONTEXT = exports.PAIRING_JSONRPC = void 0;
const time_1 = require("./time");
exports.PAIRING_JSONRPC = {
    approve: "wc_pairingApprove",
    reject: "wc_pairingReject",
    update: "wc_pairingUpdate",
    delete: "wc_pairingDelete",
    payload: "wc_pairingPayload",
    ping: "wc_pairingPing",
};
exports.PAIRING_CONTEXT = "pairing";
exports.PAIRING_DEFAULT_TTL = time_1.THIRTY_DAYS;
exports.PAIRING_SIGNAL_METHOD_URI = "uri";
exports.PAIRING_STATUS = {
    proposed: "proposed",
    responded: "responded",
    pending: "pending",
    settled: "settled",
};
exports.PAIRING_EVENTS = {
    payload: "pairing_payload",
    proposed: "pairing_proposed",
    responded: "pairing_responded",
    settled: "pairing_settled",
    updated: "pairing_updated",
    deleted: "pairing_deleted",
};
exports.PAIRING_REASONS = {
    settled: "Pairing settled",
    not_approved: "Pairing not approved",
    responded: "Pairing proposal responded",
    acknowledged: "Pairing response acknowledged",
};
